### PR TITLE
PMM-6130 Basic API for Kubernetes clusters

### PR DESCRIPTION
**Pull request by BupycHuk**
https://github.com/percona/pmm-managed/pull/440
https://github.com/percona/pmm/pull/569
https://github.com/Percona-Lab/pmm-submodules/pull/1048
https://github.com/Percona-Lab/pmm-api-tests/pull/121

### CODE REVIEW

File: models/agent_helpers.go [hunk 1]
```diff
@@ -472,7 +472,7 @@ func RemoveAgent(q *reform.Querier, id string, mode RemoveMode) (*Agent, error)
 	}
 
 	if err = q.Delete(a); err != nil {
-		return nil, errors.Wrap(q.Delete(a), "failed to delete Agent")
+		return nil, errors.Wrap(err, "failed to delete Agent")
```
Comment (AlekSi) [inline models/agent_helpers.go:475]: WHOA

File: models/kubernetes_helpers.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+func checkUniqueKubernetesClusterID(q *reform.Querier, id string) error {
+	if id == "" {
+		return status.Error(codes.InvalidArgument, "empty Kubernetes Cluster ID")
+	}
+
+	cluster := &KubernetesCluster{ID: id}
+	switch err := q.Reload(cluster); err {
+	case nil:
+		return status.Errorf(codes.AlreadyExists, "Cluster with ID %q already exists.", id)
```
Comment (AlekSi) [inline models/kubernetes_helpers.go:35]: "Kubernetes Cluster", not just "Cluster"

File: models/kubernetes_helpers.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+func checkUniqueKubernetesClusterName(q *reform.Querier, name string) error {
+	if name == "" {
+		return status.Error(codes.InvalidArgument, "empty Kubernetes Cluster Name.")
+	}
+
+	switch _, err := q.FindOneFrom(KubernetesClusterTable, "kubernetes_cluster_name", name); err {
+	case nil:
+		return status.Errorf(codes.AlreadyExists, "Cluster with Name %q already exists.", name)
```
Comment (AlekSi) [inline models/kubernetes_helpers.go:50]: "Kubernetes Cluster", not just "Cluster"

File: models/kubernetes_helpers.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	case reform.ErrNoRows:
+		return nil
+	default:
+		return errors.WithStack(err)
+	}
+}
+
+// FindAllKubernetesClusters returns all kubernetes clusters.
```
Comment (AlekSi) [inline models/kubernetes_helpers.go:58]: Kubernetes, with capital K

File: models/kubernetes_helpers.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+	clusters := make([]*KubernetesCluster, len(structs))
+	for i, s := range structs {
+		clusters[i] = s.(*KubernetesCluster)
+	}
+
+	return clusters, nil
+}
+
+// FindKubernetesClusterByName finds a kubernetes cluster with provided name.
```
Comment (AlekSi) [inline models/kubernetes_helpers.go:73]: Kubernetes, with capital K

File: models/kubernetes_helpers.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+func FindKubernetesClusterByName(q *reform.Querier, name string) (*KubernetesCluster, error) {
+	if name == "" {
+		return nil, status.Error(codes.InvalidArgument, "Empty Kubernetes Cluster Name.")
+	}
+
+	switch cluster, err := q.FindOneFrom(KubernetesClusterTable, "kubernetes_cluster_name", name); err {
+	case nil:
+		return cluster.(*KubernetesCluster), nil
+	case reform.ErrNoRows:
+		return nil, status.Errorf(codes.NotFound, "Cluster with name %q not found.", name)
```
Comment (AlekSi) [inline models/kubernetes_helpers.go:83]: "Kubernetes Cluster", not just "Cluster"

Please check all other places

File: models/kubernetes_helpers.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	default:
+		return nil, errors.WithStack(err)
+	}
+}
+
+// CreateKubernetesClusterParams contains all params required to create kubernetes cluster.
```
Comment (AlekSi) [inline models/kubernetes_helpers.go:89]: Kubernetes, with capital K

Please check all other places

File: models/kubernetes_helpers.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+type CreateKubernetesClusterParams struct {
+	KubernetesClusterName string
+	KubeConfig            string
+}
+
+// CreateKubernetesCluster creates kubernetes cluster with provided params.
+func CreateKubernetesCluster(q *reform.Querier, params CreateKubernetesClusterParams) (*KubernetesCluster, error) {
```
Comment (AlekSi) [inline models/kubernetes_helpers.go:96]: Accept pointer for consistency?

File: models/kubernetes_helpers_test.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/percona/pmm-managed/utils/testdb"
+	"github.com/percona/pmm-managed/utils/tests"
+)
+
+func TestKubernetesHelpers(t *testing.T) {
+	now, origNowF := models.Now(), models.Now
+	models.Now = func() time.Time {
+		return now
+	}
+	sqlDB := testdb.Open(t, models.SetupFixtures, nil)
```
Comment (AlekSi) [inline models/kubernetes_helpers_test.go:40]: I think `testdb.Open` already uses t.Cleanup. We can avoid adding `require.NoError(t, sqlDB.Close())` below in a new code

File: models/kubernetes_models.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+func (s *KubernetesCluster) AfterFind() error { //nolint:unparam
+	s.CreatedAt = s.CreatedAt.UTC()
+	s.UpdatedAt = s.UpdatedAt.UTC()
+
+	return nil
+}
+
+// check interfaces.
+var (
+	_ reform.BeforeInserter = (*ActionResult)(nil)
```
Comment (AlekSi) [inline models/kubernetes_models.go:63]: ⚠️ Not ActionResult

File: services/management/dbaas/kubernetes_server_test.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		ctx = logger.Set(context.Background(), t.Name())
+		uuid.SetRand(new(tests.IDReader))
+
+		sqlDB := testdb.Open(t, models.SetupFixtures, nil)
+		db = reform.NewDB(sqlDB, postgresql.Dialect, reform.NewPrintfLogger(t.Logf))
+
+		teardown = func(t *testing.T) {
+			uuid.SetRand(nil)
+
+			require.NoError(t, sqlDB.Close())
```
Comment (AlekSi) [inline services/management/dbaas/kubernetes_server_test.go:49]: Already done in t.Cleanup

**Review by AlekSi — APPROVED** [review]

**Review by AlekSi — CHANGES_REQUESTED** [review]
CI fails.

U DO NOT RUN TESTS LOCALLY? 😆

**Comment by BupycHuk** [comment]
I do, but I have some failing tests locally because of network connection and I lost these two among them.

**Review by JiriCtvrtka — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]

**Pull request merged**